packages/core/src/components/fields/NumberField.tsx

```diff
@@ -71,7 +71,7 @@ function NumberField<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends
     // Construct a regular expression that checks for a string that consists
     // of the formData value suffixed with zero or one '.' characters and zero
     // or more '0' characters
-    const re = new RegExp(`${value}`.replace('.', '\\.') + '\\.?0*$');
+    const re = new RegExp(`^(${String(value).replace('.', '\\.')})?\\.?0*$`);
```

Comment:
* Add `^` so the regular expression checks the entire string. In this bug, the value `1` matched `'231'` because this was missing.
* Wrap value in capture group and add `?` since the string representation of the value may not be present for strings like `'.001'`. This prevents the above change from breaking other cases.